src/main/kotlin/sliv/tool/scene/view/FrameView.kt

```diff
@@ -1,32 +1,55 @@
 package sliv.tool.scene.view
 
 import javafx.event.EventHandler
+import javafx.scene.Group
 import javafx.scene.canvas.*
 import javafx.scene.input.MouseEvent
+import javafx.scene.paint.Color
 import sliv.tool.scene.model.*
 import tornadofx.*
 
-class FrameView(width: Double, height: Double, private val frame: VisualizationFrame) : Fragment() {
+class FrameView(width: Double, height: Double, frame: VisualizationFrame) : Group() {
+    private var landmarksViews: Map<Layer, List<LandmarkView>>? = null
     private val canvas = Canvas(width, height)
 
-    private var landmarksViews =
-        frame.landmarks.mapValues { it.value.map { landmark -> LandmarkView.create(landmark) } }
-
-    override val root = group {
-        draw()
+    init {
+        setFrame(frame)
         canvas.onMouseMoved = EventHandler { event: MouseEvent ->
             onMouseMoved(event)
         }
         add(canvas)
     }
 
+    fun setFrame(frame: VisualizationFrame) {
+        landmarksViews = frame.landmarks.mapValues {
+            it.value.map { landmark -> LandmarkView.create(landmark) }
```

Comment:
why was it moved to `init` from variable initializer?

Reply:
``landmarksViews`` should be recalculated each time when frame changes (it is no more constant because of virtualization logic). I could manage it with the property setter, but I don't want to make it's getter public.

Reply:
can you just make private getter?